### PR TITLE
Document how to change the training objective by subclassing a trainer

### PR DESCRIPTION
**Pull request by sergiopaniego**
# What does this PR do?

Adds a `Change the training objective` section to the training customization guide.

Every section in `customization.md` today customizes training by passing an argument: optimizers and schedulers, an 8-bit reference model, callbacks, evaluation metrics, mixed precision, gradient accumulation. The guide says nothing about the case where the objective itself differs, even though subclassing a trainer and overriding `compute_loss` is a pattern the library relies on throughout:

- `GKDTrainer` and `GOLDTrainer` extend `SFTTrainer`
- `GMPOTrainer`, `MiniLLMTrainer` and `GRPOWithReplayBufferTrainer` extend `GRPOTrainer`
- `XPOTrainer` and `NashMDTrainer` extend `OnlineDPOTrainer`
- `examples/sft_diffusion_gemma/sft_diffusion_gemma.py` extends `SFTTrainer` with a block-diffusion objective, entirely from an example script

Today the only traces of this in the docs are a one-line pointer in `sft_trainer.md` and a row in `example_overview.md`.

The new section documents the three override points (`compute_loss`, the config, the collator plus `_prepare_dataset`), then links to the block-diffusion example as a complete in-repo case and to one external project using the same pattern.

Note on scope: this documents subclassing a *public* trainer for a small, local delta, which is what GMPO, GKD, GOLD, XPO and NashMD do. It does not touch the self-contained trainer route on `_BaseTrainer` described in `AGENTS.md`, which remains the way to add a full new method.

Docs only, no code changes.

## Before submitting

- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## AI writing disclosure

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed.

---

> [!NOTE]
> **Low Risk**
> Documentation-only change with no runtime or API impact.
> 
> **Overview**
> Adds a **Change the training objective** section to the training customization guide (`customization.md`), filling a gap where other sections only cover config/callback-style tweaks.
> 
> The new content explains subclassing a trainer (e.g. `DPOTrainer`) and overriding `compute_loss`, with a concrete hinge-loss DPO example using `selective_log_softmax`. It also covers extending `DPOConfig` for custom hyperparameters, supplying a custom collator and skipping default dataset prep via `_prepare_dataset`, and points readers to the in-repo block-diffusion SFT example and the external Antidoom/Antislop `DPOTrainer` extension as end-to-end references.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8710f4acf2b772a1c8ac7b8f2353b16694a874a2. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: docs/source/customization.md [hunk 1]
```diff
@@ -111,3 +111,55 @@ training_args = DPOConfig(
...
+```python
+from trl import DPOTrainer
+
+
+class MyDPOTrainer(DPOTrainer):
+    def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
+        ...
+```
+
+TRL relies on this pattern internally. [`experimental.gkd.GKDTrainer`] subclasses [`SFTTrainer`] and overrides `compute_loss` to replace the cross-entropy with a generalized Jensen-Shannon divergence against a teacher model. [`experimental.gold.GOLDTrainer`] extends [`SFTTrainer`] and [`experimental.gmpo.GMPOTrainer`] extends [`GRPOTrainer`] the same way.
```
Comment (qgallouedec) [inline docs/source/customization.md]: can you please remove this? internally, we're actually trying to do the opposite internally.

What you can say instead is that it's the simplest way to customize training, something like this.

File: docs/source/customization.md [hunk 2]
```diff
@@ -111,3 +111,55 @@ training_args = DPOConfig(
...
+
+from trl import DPOConfig
+
+
+@dataclass
+class MyDPOConfig(DPOConfig):
+    my_coef: float = field(default=0.1, metadata={"help": "Coefficient of the custom term."})
+```
+
+[`experimental.gkd.GKDConfig`] extends [`SFTConfig`] and [`experimental.gmpo.GMPOConfig`] extends [`GRPOConfig`] in the same way.
```
Comment (qgallouedec) [inline docs/source/customization.md]: same here, please remove this line

File: docs/source/customization.md [hunk 3]
```diff
@@ -111,3 +111,55 @@ training_args = DPOConfig(
...
+
+### Change the batch format
+
+When the loss needs inputs that the default collator doesn't produce, replace the collator as well, and override `_prepare_dataset` to pass the dataset through when it is already in the expected format:
+
+```python
+class MyDPOTrainer(DPOTrainer):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.data_collator = my_collator
```
Comment (qgallouedec) [inline docs/source/customization.md]: `data_collator` is an argument of `DPOTrainer`, we shouldn't need to override the init.

File: docs/source/customization.md [hunk 4]
```diff
@@ -111,3 +111,55 @@ training_args = DPOConfig(
...
+
+The loss of a trainer is defined in its `compute_loss` method. To train with a different objective, subclass the trainer and override it. Data preparation, generation, logging, and checkpointing are inherited, so the subclass holds only what actually changes.
+
+```python
+from trl import DPOTrainer
+
+
+class MyDPOTrainer(DPOTrainer):
+    def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
+        ...
```
Comment (qgallouedec) [inline docs/source/customization.md]:
```suggestion
import torch

from trl import DPOTrainer
from trl.trainer.utils import selective_log_softmax


class MyDPOTrainer(DPOTrainer):
    def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
        # In this example: the hinge loss from https://huggingface.co/papers/2309.06657
        input_ids, attention_mask = inputs["input_ids"], inputs["attention_mask"]
        shift_labels, shift_completion_mask = input_ids[:, 1:], inputs["completion_mask"][:, 1:]
        logits = model(input_ids=input_ids, attention_mask=attention_mask).logits
        with torch.no_grad():
            ref_logits = self.ref_model(input_ids=input_ids, attention_mask=attention_mask).logits

        # Sum the log-probs over the completion tokens, for the policy and for the reference model
        logps = (selective_log_softmax(logits[:, :-1], shift_labels) * shift_completion_mask).sum(dim=1)
        ref_logps = (selective_log_softmax(ref_logits[:, :-1], shift_labels) * shift_completion_mask).sum(dim=1)

        chosen_logps, rejected_logps = logps.chunk(2, dim=0)  # batch is [chosen, rejected]
        ref_chosen_logps, ref_rejected_logps = ref_logps.chunk(2, dim=0)
        delta = (chosen_logps - ref_chosen_logps) - (rejected_logps - ref_rejected_logps)
        return torch.relu(1 - self.beta * delta).mean()
```

**Review by qgallouedec — COMMENTED** [review]
super useful, thanks, a few modifications needed

**Comment by sergiopaniego** [comment]
updated, thanks for the review! @qgallouedec

**Review by qgallouedec — APPROVED** [review]
thanks, all good now.
